Mathlib/Data/Set/Defs.lean

```diff
@@ -0,0 +1,360 @@
+- `Set.diagonal`: the diagonal in `α × α`;
+- `Set.offDiag s`: the part of `s ×ˢ s` that is off the diagonal;
+- `Set.pi`: indexed product of a family of sets `∀ i, Set (α i)`,
+  as a set in `∀ i, α i`;
+- `Set.EqOn f g s`: the predicate saying that two functions are equal on a set;
+- `Set.MapsTo f s t`: the predicate syaing that `f` sends all points of `s` to `t;
+- `Set.MapsTo.restrict`: restrict `f : α → β` to `f' : s → t` provided that `Set.MapsTo f s t`;
+- `Set.restrictPreimage`: restrict `f : α → β` to `f' : (f ⁻¹' t) → t`;
+- `Set.InjOn`: the predicate saying that `f` is injective on a set;
+- `Set.SurjOn f s t`: the prediate saying that `t ⊆ f '' s`;
+- `Set.BijOn f s t`: the predicate saying that `f` is injective on `s` and `f '' s = t`;
+- `Set.graphOn`: the graph of a function on a set;
+- `Set.LeftInvOn`, `Set.RightInvOn`, `Set.InvOn`:
+  the predicates saying that `f'` is a left, right or two-sided inverse of `f` on `s`, `t`, or both;
+- `Set.image2`: the image of a pair of sets under a binary operation,
+  mostly useful to define pointwise algebraic operations on sets;
+- `Set.seq`: monadic `seq` operation on sets;
+  we don't use monadic notation to ensure support for maps between different universes;
+
+## Notations
+
+- `f '' s`: image of a set;
+- `f ⁻¹' s`: preimage of a set;
+- `s ×ˢ t`: the product of sets;
+- `s ∪ t`: the union of two sets;
+- `s ∩ t`: the intersection of two sets;
+- `sᶜ`: the complement of a set;
+- `s \ t`: the difference of two sets.
+
+## Keywords
+
+set, image, preimage
+-/
+
+-- https://github.com/leanprover/lean4/issues/2096
+compile_def% Union.union
+compile_def% Inter.inter
+compile_def% SDiff.sdiff
+compile_def% HasCompl.compl
+compile_def% EmptyCollection.emptyCollection
+compile_def% Insert.insert
+compile_def% Singleton.singleton
+
+universe u v w
+
+namespace Set
+
+variable {α : Type u} {β : Type v} {γ : Type w}
+
+@[simp, mfld_simps] theorem mem_setOf_eq {x : α} {p : α → Prop} : (x ∈ {y | p y}) = p x := rfl
+#align set.mem_set_of_eq Set.mem_setOf_eq
+
+@[simp, mfld_simps] theorem mem_univ (x : α) : x ∈ @univ α := trivial
+#align set.mem_univ Set.mem_univ
+
+instance : HasCompl (Set α) := ⟨fun s ↦ {x | x ∉ s}⟩
+
+@[simp] theorem mem_compl_iff (s : Set α) (x : α) : x ∈ sᶜ ↔ x ∉ s := Iff.rfl
+#align set.mem_compl_iff Set.mem_compl_iff
+
+instance : SDiff (Set α) := ⟨fun s t => { x | x ∈ s ∧ x ∉ t }⟩
+
+theorem diff_eq (s t : Set α) : s \ t = s ∩ tᶜ := rfl
+#align set.diff_eq Set.diff_eq
+
+@[simp] theorem mem_diff {s t : Set α} (x : α) : x ∈ s \ t ↔ x ∈ s ∧ x ∉ t := Iff.rfl
+#align set.mem_diff Set.mem_diff
+
+theorem mem_diff_of_mem {s t : Set α} {x : α} (h1 : x ∈ s) (h2 : x ∉ t) : x ∈ s \ t := ⟨h1, h2⟩
+#align set.mem_diff_of_mem Set.mem_diff_of_mem
+
+-- Porting note: I've introduced this abbreviation, with the `@[coe]` attribute,
+-- so that `norm_cast` has something to index on.
+-- It is currently an abbreviation so that instance coming from `Subtype` are available.
```

Comment:
This porting note should be updated, as you've now made it a reducible definition: clearly it hasn't broken anything but I think we shouldn't leave this in in the current form.

Reply:
I copy+pasted it. Is there any difference between an `abbrev` and a `@[reducible] def` in Lean 4?

Reply:
I thought they were synonyms.

Reply:
I have the same impression.

Reply:
Oh, grand. I am a little unclear on the differences between these things.

Reply:
There is at least a minor [difference](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/changing.20reducibility.20for.20type.20class.20inference/near/411940454): you can't write `abbrev` on a recursive definition.

Further `abbrev` [adds](https://github.com/leanprover/lean4/blob/b0fe9d6cdca82723134d61f8fe23a8c674f308b7/src/Lean/Elab/DefView.lean#L47-L53) `@[inline]`.

Reply:
I think it's fine to leave the comment, as it's orthogonal to this moving-things-around PR, but it would also be fine to update the comment replacing `abbrev` everywhere with "reducible def".